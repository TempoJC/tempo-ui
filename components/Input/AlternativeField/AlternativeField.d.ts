export default AlternativeField;
/**
 * Base component for defining an alternative form field component.
 * This is similar to `Field`, but it is designed for rendering borderless inputs.
 */
declare function AlternativeField({ children, dense, disabled, readOnly, label, color, onChange, onBlur, onFocus, required, errors, name, className, id, helper, testId, tabIndexValues, ...rest }: {
    [x: string]: any;
    children: any;
    dense?: boolean | undefined;
    disabled?: boolean | undefined;
    readOnly?: boolean | undefined;
    label: any;
    color?: string | undefined;
    onChange?: (() => void) | undefined;
    onBlur?: (() => void) | undefined;
    onFocus?: (() => void) | undefined;
    required?: boolean | undefined;
    errors?: never[] | undefined;
    name: any;
    className: any;
    id?: any;
    helper: any;
    testId: any;
    tabIndexValues?: {} | undefined;
}): import("react/jsx-runtime").JSX.Element;
