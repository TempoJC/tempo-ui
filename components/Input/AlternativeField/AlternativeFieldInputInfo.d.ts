export default InputInfo;
declare function InputInfo({ testId, className, disabled, helper, errors }: {
    testId: any;
    className: any;
    disabled: any;
    helper: any;
    errors?: never[] | undefined;
}): import("react/jsx-runtime").JSX.Element | null;
