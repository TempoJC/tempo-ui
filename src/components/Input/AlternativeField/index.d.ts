import type { ChangeEventHandler, FocusEventHandler, ReactElement, ReactNode } from "react";
import { CommonProps } from "@/types";

export type AlternativeFieldFuncProps<Value> = CommonProps & {
  dense: boolean;
  disabled: boolean;
  hasError: boolean;
  id?: string;
  name?: string;
  onBlur: FocusEventHandler;
  onChange: ChangeEventHandler;
  onFocus: FocusEventHandler;
  readOnly: boolean;
  setHasError: (error: boolean) => void;
  value?: Value;
};

export type AlternativeFieldCommonProps = {
  dense?: boolean;
  disabled?: boolean;
  errors?: ReactNode[];
  helper?: ReactNode;
  id?: string;
  label?: ReactNode;
  name?: string;
  readOnly?: boolean;
  required?: boolean;
};

export type AlternativeFieldProps<Value, Input = unknown> = CommonProps &
  AlternativeFieldCommonProps & {
    children?: (props: AlternativeFieldFuncProps<Value>) => ReactElement;
    onBlur?: FocusEventHandler<Input>;
    onChange?: ChangeEventHandler<Input>;
    onFocus?: FocusEventHandler<Input>;
    value?: Value;
  };

declare function AlternativeField<Value>(props: AlternativeFieldProps<Value>): JSX.Element;

export default AlternativeField;
