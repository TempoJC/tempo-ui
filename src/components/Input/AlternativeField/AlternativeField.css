.tempo-alternative-field__input {
  display: flex;
  align-items: center;
}

@define-mixin required-alternative-field-label {
  .tempo-alternative-field--required &:after {
    content: "*";
    display: inline-block;
    padding-left: var(--size1);
    text-align: left;
    vertical-align: top;
  }
}

.tempo-alternative-field__label {
  display: block;
  color: var(--color-token-light-content-3);
  font: var(--typography-detail-m-regular-font);
  letter-spacing: var(--typography-detail-m-regular-letter-spacing);
  margin-bottom: calc(var(--size2) + 1px);

  .tempo-alternative-field--color-dark & {
    color: var(--color-token-dark-content-3);
  }

  @mixin required-alternative-field-label;
}

.tempo-alternative-field--dense .tempo-alternative-field__label {
  margin-bottom: 0;
}

@media (--viewport-xs-m) {
  .tempo-alternative-field__label {
    font: var(--typography-body-m-regular-font);
    letter-spacing: var(--typography-body-m-regular-letter-spacing);
  }
}

.tempo-alternative-field__label--has-error {
  color: var(--color-token-light-semantic-danger);

  .tempo-alternative-field--color-dark & {
    color: var(--color-token-dark-semantic-danger);
  }
}

.tempo-alternative-field__label--disabled {
  color: var(--color-token-light-content-disabled);

  .tempo-alternative-field--color-dark & {
    color: var(--color-token-dark-content-disabled);
  }
}
