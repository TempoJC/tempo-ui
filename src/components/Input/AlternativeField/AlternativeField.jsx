import { useState } from "react";
import cn from "classnames";
import InputInfo from "./AlternativeFieldInputInfo.jsx";
import { TabIndexContextProvider } from "@/hooks/index.js";
import { is, noop } from "@/utils";
import "./AlternativeField.css";

/**
 * Base component for defining an alternative form field component.
 * This is similar to `Field`, but it is designed for rendering borderless inputs.
 */
const AlternativeField = ({
  children,
  dense = false,
  disabled = false,
  readOnly = false,
  label,
  color = "light",
  onChange = noop,
  onBlur = noop,
  onFocus = noop,
  required = false,
  errors = [],
  name,
  className,
  id = name,
  helper,
  testId,
  tabIndexValues = {},
  ...rest
}) => {
  const [hasError, setHasError] = useState(errors.length > 0);

  const classNames = cn("tempo-alternative-field", className, {
    "tempo-alternative-field--dense": dense,
    "tempo-alternative-field--required": required,
    "tempo-alternative-field--color-dark": color === "dark",
  });

  const inputProps = {
    ...rest,
    className: "tempo-alternative-field-input",
    disabled,
    readOnly,
    id,
    name,
    onBlur,
    onFocus,
    onChange,
    hasError,
    setHasError,
    color,
    testId,
  };

  return (
    <TabIndexContextProvider value={tabIndexValues}>
      <div className={classNames} data-testid={testId}>
        {label && (
          <label
            htmlFor={id}
            className={cn("tempo-alternative-field__label", {
              "tempo-alternative-field__label--has-error": hasError,
              "tempo-alternative-field__label--disabled": disabled,
            })}
            data-testid={testId && `${testId}-label`}
          >
            {label}
          </label>
        )}
        <div className="tempo-alternative-field__input">{is(Function, children) && children(inputProps)}</div>
        <InputInfo testId={testId && `${testId}-inputInfo`} helper={helper} errors={errors} disabled={disabled} />
      </div>
    </TabIndexContextProvider>
  );
};

export default AlternativeField;
