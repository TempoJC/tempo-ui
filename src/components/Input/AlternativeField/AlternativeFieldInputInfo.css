.tempo-alternative-field__helper {
  margin-top: var(--size2);
  color: var(--color-token-light-content-1);
  font: var(--typography-detail-m-light-font);
  letter-spacing: var(--typography-detail-m-light-letter-spacing);

  .tempo-alternative-field--color-dark & {
    color: var(--color-token-dark-content-1);
  }
}

.tempo-alternative-field--dense .tempo-alternative-field__helper {
  margin-top: 0;
}

@media (--viewport-xs-m) {
  .tempo-alternative-field__helper {
    font: var(--typography-detail-m-light-font);
    letter-spacing: var(--typography-detail-m-light-letter-spacing);
  }
}

.tempo-alternative-field__helper--disabled {
  color: var(--color-token-light-content-disabled);

  .tempo-alternative-field--color-dark & {
    color: var(--color-token-dark-content-disabled);
  }
}

.tempo-alternative-field__errors {
  margin-top: var(--size2);
  list-style: none;
  color: var(--color-token-light-semantic-danger);
  font: var(--typography-detail-m-light-font);
  letter-spacing: var(--typography-detail-m-light-letter-spacing);

  .tempo-alternative-field--color-dark & {
    color: var(--color-token-dark-semantic-danger);
  }
}

.tempo-alternative-field--dense .tempo-alternative-field__errors {
  margin-top: 0;
}

@media (--viewport-xs-m) {
  .tempo-alternative-field__errors {
    font: var(--typography-detail-m-light-font);
    letter-spacing: var(--typography-detail-m-light-letter-spacing);
  }
}

.tempo-alternative-field__errors--disabled {
  color: var(--color-token-light-content-disabled);

  .tempo-alternative-field--color-dark & {
    color: var(--color-token-dark-content-disabled);
  }
}
