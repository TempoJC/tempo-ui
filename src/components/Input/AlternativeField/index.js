export { default as AlternativeField } from "./AlternativeField";
