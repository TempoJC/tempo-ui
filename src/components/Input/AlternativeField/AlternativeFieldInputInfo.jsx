import cn from "classnames";
import "./AlternativeFieldInputInfo.css";

const InputHelper = ({ testId, className, disabled, message }) => (
  <div
    className={cn(
      "tempo-alternative-field__helper",
      {
        "tempo-alternative-field__errors--disabled": disabled,
      },
      className,
    )}
    data-testid={testId}
  >
    {message}
  </div>
);

const InputErrors = ({ testId, className, disabled, errors }) => {
  if (!errors.length) {
    return null;
  }

  return (
    <ul
      className={cn(
        "tempo-alternative-field__errors",
        {
          "tempo-alternative-field__errors--disabled": disabled,
        },
        className,
      )}
      data-testid={testId}
    >
      {errors.map((error, i) => (
        <li key={i} data-testid={testId && `${testId}-${i}`}>
          {error}
        </li>
      ))}
    </ul>
  );
};

const InputInfo = ({ testId, className, disabled, helper, errors = [] }) => {
  if (errors.length > 0) {
    return (
      <InputErrors testId={testId && `${testId}-errors`} className={className} disabled={disabled} errors={errors} />
    );
  }

  return helper ? (
    <InputHelper testId={testId && `${testId}-helper`} className={className} disabled={disabled} message={helper} />
  ) : null;
};

export default InputInfo;
