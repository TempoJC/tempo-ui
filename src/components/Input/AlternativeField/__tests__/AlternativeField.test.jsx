import { render, screen, getAllByText } from "@testing-library/react";
import AlternativeField from "../AlternativeField";

describe("@tempo-ui/Input/AlternativeField", () => {
  it("renders", () => {
    const { container } = render(<AlternativeField />);
    expect(container).toMatchSnapshot();
  });

  it("renders a label", () => {
    render(<AlternativeField label="Label" />);
    expect(screen.getByText("Label")).toBeInTheDocument();
  });

  it("renders a helper text", () => {
    render(<AlternativeField helper="Helper Text" />);
    expect(screen.getByText("Helper Text")).toBeInTheDocument();
  });

  it("renders a single line error text, inside a list", () => {
    render(<AlternativeField errors={["Error1"]} testId="testId" />);

    const errorsComponent = screen.getByTestId("testId-inputInfo-errors");
    const errors = getAllByText(
      errorsComponent,
      (content, element) => content === "Error1" && element.tagName.toLowerCase() === "li",
    );
    expect(errors).toHaveLength(1);
  });

  it("renders list of multiple errors", () => {
    render(<AlternativeField errors={["Error1", "Error2", "Error3"]} testId="testId" />);

    const errorsComponent = screen.getByTestId("testId-inputInfo-errors");
    const errors = getAllByText(
      errorsComponent,
      (content, element) => content.match(/Error[0-3]/) && element.tagName.toLowerCase() === "li",
    );

    expect(errors).toHaveLength(3);
  });

  it("adds the additional className", () => {
    const { container } = render(<AlternativeField className="customClass" />);

    expect(container.querySelector(".customClass")).toBeInTheDocument();
  });

  describe("testId prop", () => {
    it("renders the alternativeButton with the specified data-testid attribute", () => {
      const { getByTestId } = render(<AlternativeField testId="alternative-field" />);

      const element = getByTestId("alternative-field");

      // Verifies that the element is in the document (implicit by getByTestId, but good practice)
      expect(element).toBeInTheDocument();

      // Optional: If you want to be even more explicit about the attribute
      expect(element).toHaveAttribute("data-testid", "alternative-field");
    });
  });

  describe("dense", () => {
    it("renders a dense alternative field", () => {
      const { container } = render(<AlternativeField dense />);

      expect(container.querySelector(".tempo-alternative-field--dense")).toBeInTheDocument();
    });
  });
});
