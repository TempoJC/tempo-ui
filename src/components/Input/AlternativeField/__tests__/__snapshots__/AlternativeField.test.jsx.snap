// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`@tempo-ui/Input/AlternativeField renders 1`] = `
<div>
  <div
    class="tempo-alternative-field"
  >
    <div
      class="tempo-alternative-field__input"
    />
  </div>
</div>
`;
